src/libcore/fmt/float.rs

```diff
@@ -23,15 +50,38 @@ fn float_to_decimal_common<T>(fmt: &mut Formatter, num: &T, negative_zero: bool)
         (true,  true)  => flt2dec::Sign::MinusPlusRaw,
     };
 
-    let mut buf = [0; 1024]; // enough for f32 and f64
-    let mut parts = [flt2dec::Part::Zero(0); 16];
-    let formatted = if let Some(precision) = fmt.precision {
-        flt2dec::to_exact_fixed_str(flt2dec::strategy::grisu::format_exact, *num, sign,
-                                    precision, false, &mut buf, &mut parts)
+    if let Some(precision) = fmt.precision {
+        float_to_decimal_common_exact(fmt, num, sign, precision)
     } else {
-        flt2dec::to_shortest_str(flt2dec::strategy::grisu::format_shortest, *num, sign,
-                                 0, false, &mut buf, &mut parts)
-    };
+        float_to_decimal_common_shortest(fmt, num, sign)
+    }
+}
+
+// Don't inline this so callers don't use the stack space this function
+// requires unless they have to.
+#[inline(never)]
+fn float_to_exponential_common_exact<T>(fmt: &mut Formatter, num: &T,
+                                        sign: flt2dec::Sign, precision: usize,
+                                        upper: bool) -> Result
+    where T: flt2dec::DecodableFloat
+{
+    let mut buf: [u8; 1024] = unsafe { mem::uninitialized() }; // enough for f32 and f64
+    let mut parts: [flt2dec::Part; 7] = unsafe { mem::uninitialized() };
+    let formatted = flt2dec::to_exact_exp_str(flt2dec::strategy::grisu::format_exact,
+                                              *num, sign, precision,
+                                              upper, &mut buf, &mut parts);
+    fmt.pad_formatted_parts(&formatted)
+}
+
+fn float_to_exponential_common_shortest<T>(fmt: &mut Formatter,
```

Comment:
This should also be a `#[inline(never)]`, even though `MAX_SIG_DIGITS` is only `17`. This is because when this gets inlined into `float_to_exponential_common` and formatting eventually calls `float_to_exponential_common_exact`, it results in combined stack usage equal to the sum of stack usage of `float_to_exponential_common_shortest` and `float_to_exponential_common_exact`.